corrections on TOSS STIG items in the 010000's

sbalingit211 opened the pull request:
Made corrections for TOSS STIG items in the 010000's as needed. Some did not do all the asks in the checktext, while others may prevent the ansible playbook from completing or may break a system or cluster.

sbalingit211 committed:
corrections on items that did not do the full checktext ask or may break a system or ansible playbook

sbalingit211 committed:
add ignore_errors to tasks with assert modules and correct value of ignore_errors

sbalingit211 committed:
remove toss_stig tags

IanLee1521 commented on .ansible-lint-ignore:
```diff
@@ -8,3 +8,9 @@ ansible/roles/stig/tasks/TOSS_04_020160.yml fqcn[action-core]
 ansible/roles/stig/tasks/TOSS_04_030900.yml command-instead-of-shell
 ansible/roles/stig/tasks/TOSS_04_030900.yml risky-shell-pipe
+
+# below rules ignore "failed_when" so when tasks fail, the ansible playbook will continue
```
I'm not sure I understand why these rules are ignoring errors.

sbalingit211 replied:
([copying from mattermost post](https://mattermost.energy.gov/lc-llnl/pl/5kha5fb75tge8bagzd3f138iyc))

I spoke with @cdefrates and we agree that the lesser evil would be to have the `ignore_errors: true` for these specific needs, at least until we find a better method. One I had are to change them to `ansible.builtin.debug` so it prints out the failed statement, but the task result shows up as "ok". Another idea I had was to add a final task to each `TOSS-04-*.yml` file that says whether said stig passed or not, but I'm worried about runtime considering that means adding 227 more tasks AND having to register a variable to all subtasks that we care about to justify whether a toss stig item passed or not.

For details on our tests on `failed_when: ${some_conditional}`, `failed_when: false`, `ignore_errors: true`, and `quiet: true`, see below:


- for items that have some ansible task with `failed_when: ${some conditional}` and the ansible task failed, the playbook stops. For example, the task is:
```
      ansible.builtin.command: "grep hosts /etc/nsswitch.conf"
      changed_when: false
      check_mode: false
      failed_when: dns_entry.rc not in [0,1]
      register: dns_entry
```
.
		- when `/etc/nsswitch.conf` does not exist, the playbook ends:
```
TASK [toss_stig : TOSS-04-010330 - Determine whether the system is using local or DNS name resolution] ************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "cmd": ["grep", "hosts", "/etc/nsswitch.conf"], "delta": "0:00:00.008253", "end": "2024-12-10 13:51:06.446974", "failed_when_result": true, "msg": "non-zero return code", "rc": 2, "start": "2024-12-10 13:51:06.438721", "stderr": "grep: /etc/nsswitch.conf: No such file or directory", "stderr_lines": ["grep: /etc/nsswitch.conf: No such file or directory"], "stdout": "", "stdout_lines": []}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```
.
	- for the same thing, but the `failed_when` conditional is met, it stops the playbook from completing:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "resolv_stat.stat.size == 0", "changed": false, "evaluated_to": false, "msg": "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```
.
- if I add a `failed_when: false` to a task that has a `failed_when: ${some conditional}`, a warning appears AND the task shows up as okay even though it should have failed:
```
TASK [Check for incompatible operating systems] *******************************************************************************************************************************************
ok: [localhost]
**[WARNING]: While constructing a mapping from /etc/ansible/roles/toss_stig/tasks/TOSS_04_010330.yml, line 20, column 7, found a duplicate dict key (failed_when). Using last defined value**
**only.**
included: /etc/ansible/roles/toss_stig/tasks/TOSS_04_010330.yml for localhost

TASK [toss_stig : TOSS-04-010330 - Determine whether the system is using local or DNS name resolution] ************************************************************************************
ok: [localhost]
```
.
- if I remove the conditional and only have `failed_when: false`, the task shows up as "ok". The below example is when `/etc/nsswitch.conf` does not exist so it should show up with an error message that the file does not exist:
```
TASK [toss_stig : TOSS-04-010330 - Determine whether the system is using local or DNS name resolution] ************************************************************************************
ok: [localhost]
```
.
- for items that have use `ansible.builtin.assert` to only fail based on a conditional and has `quiet:true`, the playbook stops running. For example, say I have this task:
```
    - name: TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty
      ansible.builtin.assert:
        that:
          - resolv_stat.stat.size == 0
        fail_msg: "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"
        quiet: true
```
.
		- the playbook shows the fatal message and stops running:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "resolv_stat.stat.size == 0", "changed": false, "evaluated_to": false, "msg": "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"}

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=1    skipped=1    rescued=0    ignored=0
```
.
- say I add `failed_when: false` to the task, the task results in "ok" nor does it print the fail message wanted, and continues with the playbook:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
ok: [localhost]

TASK [toss_stig : TOSS-04-010330 - Verify the operating system is configured to use two or more name servers for DNS resolution] **********************************************************
skipping: [localhost]
```
.
	- if `ignore_errors: true` is added, the task results show "failed" and the error message and continues the playbook, but `...ignoring` follows the message and the recap counts it as a "skipped" task:
```
TASK [toss_stig : TOSS-04-010330 - Verify that "/etc/resolv.conf" is empty] ***************************************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "resolv_stat.stat.size == 0", "changed": false, "evaluated_to": false, "msg": "If local host authentication is being used, the /etc/resolv.conf file must be emtpy"}
...ignoring

TASK [toss_stig : TOSS-04-010330 - Verify the operating system is configured to use two or more name servers for DNS resolution] **********************************************************
skipping: [localhost]

TASK [toss_stig : TOSS-04-010330 - Check that there are at least two name servers for DNS resolution] *************************************************************************************
fatal: [localhost]: FAILED! => {"assertion": "num_name_servers.stdout is defined", "changed": false, "evaluated_to": false, "msg": "Configure the operating system to use two or more name servers for DNS resolution"}
...ignoring

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0    skipped=2    rescued=0    **ignored=2**
```

IanLee1521 replied:
ok

IanLee1521 commented on ansible/roles/stig/tasks/TOSS_04_010110.yml:
```diff
@@ -15,15 +15,25 @@
-  notify: Reload sysctl
+  block:
+    - name: TOSS-04-010110 - check files if correct setting exists in /etc/sysctl.d/
+      ansible.builtin.find:
```
We might want to consider switching these to use the `ansible.posix.sysctl` module in the future: https://docs.ansible.com/ansible/latest/collections/ansible/posix/sysctl_module.html

Not needed for this MR though.

sbalingit211 replied:
sounds good, adding it to my list of future features to investigate at a later time

IanLee1521 replied:
Is there an issue created for that (here in this repo) to track?

sbalingit211 replied:
Issue has been created: https://github.com/LLNL/toss-stig/issues/65

IanLee1521 commented on ansible/roles/stig/tasks/TOSS_04_010330.yml:
```diff
@@ -30,42 +30,78 @@
+    - toss_04_010330 | bool
+  tags:
+    - toss_stig
+- name: TOSS-04-010330 - fail if either nsswitch or resolv file does not exist
```
GitHub seems to agree with my questions about using the ignore_errors ?

sbalingit211 replied:
see above for explanation.

IanLee1521 commented on ansible/roles/stig/tasks/TOSS_04_010430.yml:
```diff
@@ -25,23 +14,38 @@
-    mode: '0644'
+  block:
+    - name: TOSS-04-010430 - check if gdm package is installed
+      ansible.builtin.shell: "rpm -qa | grep gdm"
```
should this use the rpm / package module instead of a shell command?

sbalingit211 replied:
This has been updated to use the "package" module

IanLee1521 commented on ansible/roles/stig/tasks/TOSS_04_010370.yml:
```diff
@@ -9,21 +9,20 @@
-        file_type: file
-        recurse: false
+      # as of 2024-12-02, ansible.builtin.find does not have the capability to include the '-xdev' option
+      ansible.builtin.command: "find / -xdev -name '*.shosts'"
```
I'm not sure that using the `-xdev` flag is really what we want here... Technically we might want to check the entire system? (yes, I know that doesn't really work in our use case internally due to large systems, but as the standard answer on this STIG it might be the better answer.)

sbalingit211 replied:
This task has been reverted back to the previous version

IanLee1521 commented on ansible/roles/stig/tasks/TOSS_04_010350.yml:
```diff
@@ -25,6 +25,7 @@
           - uidzero.stdout_lines | length == 0
         fail_msg: "Another account has been found to have uid 0 in /etc/passwd besides root."
         quiet: true
+  ignore_errors: true
```
Why ignore errors here?

sbalingit211 replied:
This falls under the explanation above: when this "assert" module activates, it stops the playbook from completing. This will at least have the task show the message this task failed and move on, but will show up as "skipped" in the recap.

sbalingit211 committed:
minor changes for PR comments:
- revert back TOSS-04-010370
- use package module instead of shell module

other minor changes:
- add "manual" tag to TOSS-04-020300
- remove duplicate checktexts
